From: Alessandro Astone <alessandro.astone@canonical.com>
Date: Wed, 4 Jun 2025 18:20:51 +0200
Subject: cogl/pipeline: Make foreach_child function safe

The cogl_pipeline_foreach_child function is not safe to callbacks that modify
the list of children.

Indeed, the reparent_children_cb callback function modifies the list and causes
unexpected behaviour.

Fixes: 8540362d1d ("cogl/pipeline: Inherit from GObject")

Origin: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4481
Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/3970
Bug-Ubuntu: https://launchpad.net/bugs/2102063
---
 cogl/cogl/cogl-pipeline.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/cogl/cogl/cogl-pipeline.c b/cogl/cogl/cogl-pipeline.c
index 2e7b595..f93aaa8 100644
--- a/cogl/cogl/cogl-pipeline.c
+++ b/cogl/cogl/cogl-pipeline.c
@@ -121,11 +121,14 @@ cogl_pipeline_foreach_child (CoglPipeline              *pipeline,
 {
 
   for (CoglPipeline *child = pipeline->first_child;
-       child != NULL;
-       child = child->next_sibling)
+       child != NULL;)
     {
+      CoglPipeline *next = child->next_sibling;
+
       if (!callback (child, user_data))
         break;
+
+      child = next;
     }
 }
 
